pandas/core/computation/eval.py

```diff
@@ -232,8 +237,13 @@ def eval(expr, parser='pandas', engine=None, truediv=True,
     pandas.DataFrame.query
     pandas.DataFrame.eval
     """
-    inplace = validate_bool_kwarg(inplace, 'inplace')
-    first_expr = True
+
+    inplace = validate_bool_kwarg(inplace, "inplace")
+    modifiable = hasattr(target, "__setitem__")
+
+    if inplace and not modifiable:
+        raise ValueError("Cannot modify the provided target inplace")
```

Comment:
I think a uniform error message is needed.  Passing in `np.ndarray` provides an error message that is not immediately clear regarding item assignment.

> people should actually only pass dataframes

Well perhaps `pd.eval` should not exist. :smile:  Because as is, there's nothing wrong with me passing in a dictionary or some other dict-like object.